### ISSUE TITLE
Proposal to change the type system for Souffle to include integer types, booleans and floating point numbers

### ISSUE THREAD

**Issue by loadedice** [souffle-lang/souffle#456]
Currently Souffle does not support specifying an integers size or if it is signed or not, it also does not support floating point numbers or have a concise way to represent a boolean variable.

I propose the following changes:

+ `number` to be deprecated, issuing a warning to the user when it is used
+ Support for signed integers of different size:
    * `i64` for 64 bit signed integers
    * `i32` for 32 bit signed integers
    * `i16` for 16 bit signed integers
    * `i8` for 8 bit signed integers
+ Support for unsigned integers of different size:
    * `u64` for 64 bit unsigned integers
    * `u32` for 32 bit unsigned integers
    * `u16` for 16 bit unsigned integers
    * `u8` for 8 bit unsigned integers
+ Support of floating point numbers in the following representations:
    * `f32` for a 32 bit float (typically called `float`)
    * `f64` for a 64 bit float (typically called `double`)
+ Support for booleans:
    * `bool` for a boolean which can either be `true` or `false`

The naming convention is based off [Rusts type system](https://doc.rust-lang.org/reference/types.html#types) (although there exists C code predating Rust and `stdint.h` with typedefs similar to this), however this decision is arbitrary and up for debate, I personally like it because it is concise and familiar to some users.

When using the new types we would not allow overflows or underflows to occur, if we had a program such as this:

```
.decl R(n: u16)
.decl A(n: u16)
...
R(a + b) :- A(a), A(b).
```

We would want to discard any tuples generated if `a + b` is the result of an overflow, so an implicit `success_addition` constraint will be added to check that `a + b` does not result in an overflow for `u16` types. Furthermore we would also want to place similar constraints on all the integer types for all operators and other specific constraints for floating points. Another point of discussion would be should there be a way to disable these constraints if for instance an overflow is expected behaviour.

Some examples of using the new proposed syntax is as follows:

```
.decl odd(n: u8)

// Generate odd numbers between 1 and 255
odd(1).
odd(n + 2) :- odd(n).

```

Should output:

```
---------------
odd
===============
1
3
5
7
...
251
253
255
===============
```

Example of using booleans:

```
.type Person

.decl people(p: Person)
.decl is_happy(p: Person, happy: bool)
.decl souffle_has_better_types(b: bool)

people("Joel").
souffle_has_better_types(true).
is_happy(x, true) :- people(x), souffle_has_better_types(true).

.output is_happy
```

Should output:

```
---------------
is_happy
===============
Joel	true
===============
```

Example using floating points:

```
.decl less_than_five(n: f32)

less_than_five(0.0).
less_than_five(x + 0.5) :- less_than_five(x), (x + 0.5) < 5.0.

.output less_than_five
```

Should output:

```
---------------
less_than_five
===============
0.0
0.5
1.0
1.5
2.0
2.5
3.0
3.5
4.0
4.5
===============
```

Please let me know if you have any ideas or thoughts on this, it'd be greatly appreciated!

**Comment by lyndonhenry** [comment]
So I really like this, it's a great idea.

As a possible extension to this, as it would only really be possible by implementing the mechanics for the proposed type system already, it'd be cool to be able to define types in a similar syntax to datalog programs. Then in the compilation/interpretation have some sort of analyzer that determines the most efficient type to use. 

For example, we can define ints and doubles by the following. 

~~~
// int + float is undefined
.type i : number
i(x) :- x + 0.1 != x + 0.1.

// a 32 bit int is an int in the given range
.type i32 : number
i32(x) :- i(x), x >= -2147483648, x <= 2147483647.

// float + float is always defined
.type f : number
f(x) :- x <= 0, x + 0.1 = x + 0.1.
f(x) :- x > 0, x - 0.1 = x - 0.1.

// a 64 bit float is a float in the given range
.type f64 : number
f64(x) :- f(x), x >= 2.2E-308, x <= 1.8E+308. 

// a boolean is an int that is either 0 or 1
.type bool : number
bool(x) :- i(x), x >= 0, x <= 1.

// predicate demonstrating use of types
.decl P(x:i32, y:f64, z:bool)
P(1, 1, 1)

// this is the same as above, but should fail as a predicate and a type are different things
.decl Q(x:number, y:number, z:number)
Q(x, y, z) :- i32(x), f64(y), bool(z), x = 1, y = 1, z = 1.
~~~

This becomes quite fun when one considers the possibilites for predicate overloading and mixing different systems together.

For example, something like the following could be used for defining implication in Kleene and Lukasiewicz logic (two kinds of three valued logic in which there is true, false, and unknown). 

~~~
.type value : number
value(x) :- i(x), x >= -1, x <= 1.

.type kleene : value
.type lukasiewicz : value

.decl implication(A:value, B:value, C:value)

// kleene says that "unknown -> unknown" is "unknown"
implication(A:kleene, B:kleene, C:kleene) -: A = 0, B = 0, C = 0.

// while lukasiewicz says that "unknown -> unknown" is "true"
implication(A:lukasiewicz, B:lukasiewicz, C:lukasiewicz) -: A = 0, B = 0, C = 1.
~~~

Anyway, all just a thought for future at this point.

**Comment by mmcgr** [comment]
This sounds great, but I'm not so sure about 
> We would want to discard any tuples generated if a + b is the result of an overflow

Silent errors can be hard to diagnose. I think a better default would be to exit with an error. Is there a case where undefined behaviour - the results of an overflow - is desirable?

We do currently depend on that undefined behaviour to handle cases like -2^31, but I really don't think it's desirable, just much simpler. (e.g. -2^31 in a datalog file is actually interpreted as - (2^31), 2^31 does not exist but with undefined behaviour it becomes, on x86,  -2^31, -(-2^31) would be 2^31 but that does not exist but with undefined behaviour it becomes, on x86, -2^31. I looked into fixing this, but since souffle really only works on x86 and the result of fixing it would be identical, but slower, I let it be.)

**Comment by scyptnex** [comment]
Some things to consider re floating-point:
* epsilon differences: pretend a+b = c even if its _slightly_ off
* "NaN" and "Infty" are now valid numbers (that need to be scannable in the parser)
* In general a+b != b+a for float32s

**Comment by loadedice** [comment]
So in order to implement the sized integers we have settled on implementing a range type that only allows values within a certain range, and then aliasing these types to be specific ranges, like u8 would be [0, 255], etc

The syntax is as follows:
```
.type zerotosix: 0..6

.decl A(n:zerotosix)
A(0).
A(5).
A(6).
A(x+1) :- A(x), x < 6.

.output A
```
And if you are to make an error this happens:
```
.decl A(n: i8)

A(1000).
```
produces the following error:
```
Error: Number constant not in range [-128, 127] in file out_of_range.dl at line 3
A(1000).
--^------
1 errors generated, evaluation aborted
```

There are some choices to be made:

- To deal with values being inserted into a relation, should an error be raised or should they be silently ignored if they fall out of the range.

I have decided that we will only allow tuples from range type that is smaller (or equal) to fit into the other range types, and if a user tries to do otherwise then an error will be generated like above. An example where an error would be raised is the following:
```
.type zerotosix:0..6
.type zerototen:0..10

.decl A(x:zerototen)
A(0).
A(2).
A(4).
.output A

.decl B(x:zerotosix)
B(x) :- A(x).
.output B
```
However doing the opposite, that is, trying to fit a zerotosix into a zerototen should be fine.

Although it would be nice to have a run time error so that we can only throw the error when it actually causes an issue, it looks like the way things are evaluating in Souffle make it hard to check at the time of inserting into a relation (relations are inserted at the same time that the conditions are checked) and since we don't know what type things are at the run time it also becomes an issue.
